### PR TITLE
chore: update Nx to v16.1.4 and run migrations

### PR DESCRIPTION
**Pull request by rathpc**
**Following this merge and release, another PR will be needed to update the dependencies referencing this own packages exports as well as removing the peerDependency ignore as the older nrwl packages will no longer be transitively referenced.**

This also updates:

- nx-remotecache-custom to v4 which supports Nx v16
- ngx-deploy-npm to v6 which supports Nx v16

BREAKING CHANGES: Minimum Nx version is now >=16.0.0

### CODE REVIEW

**Comment by rathpc** [comment]
@robinpellegrims hello again! Ok this one will update nx and a couple dependencies but we need to follow this one up with 1 or two more. The first is to update `@pellegrims` dependencies to the newly released versions after this is released and also remove the peerDependency ignores. Then I will open another to update all other dependencies that I can safely bump to get everything else up to date

Please let me know if you have any questions

**Comment by rathpc** [comment]
@robinpellegrims I wonder, is it possible to actually remove the `@pellegrims` dependencies and reference those things as local packages where possible? I think having the self references are complicating things a bit

**Comment by robinpellegrims** [comment]
> @robinpellegrims I wonder, is it possible to actually remove the `@pellegrims` dependencies and reference those things as local packages where possible? I think having the self references are complicating things a bit

Yeah this is possible, although I'm not sure it's really necessary.

We can reference the nx-remotecache-s3 runner like this: `"runner": "./dist/libs/nx-remotecache-s3",`. We just need to make sure it's built prior to doing any other tasks that require the remote cache. For the eslint package it's similar.

File: apps/cotersus-e2e/project.json
Comment (robinpellegrims) [inline apps/cotersus-e2e/project.json:4]: this removal is causing the ci to fail because it cannot determine the affected projects
I'm trying to fix it through [this issue](https://github.com/robinpellegrims/pellegrims/issues/216)

**Review by robinpellegrims — CHANGES_REQUESTED** [review]
Looks good to me, but we have some conflicts now after my ci fixes.
Could you rebase on master, fix conflicts and force push?

**Comment by rathpc** [comment]
> Looks good to me, but we have some conflicts now after my ci fixes.
> Could you rebase on master, fix conflicts and force push?

Did you want the runner and lint configs referenced relative or keep it as it is?

**Comment by robinpellegrims** [comment]
> > Looks good to me, but we have some conflicts now after my ci fixes.
> > Could you rebase on master, fix conflicts and force push?
> 
> Did you want the runner and lint configs referenced relative or keep it as it is?

I would say we keep referencing the compiled packages for now

**Comment by rathpc** [comment]
@robinpellegrims just rebased and force pushed. Let me know when the new release is available and I will create the next one updating the package versions along with other dependencies

**Comment by rathpc** [comment]
looks like `e-square-io/nx-distributed-task@v2` is expecting `@nrwl/cli` to be available but with this update that scope changes to `@nx/cli` we might need an update in that repo as well with a major bump supporting nx 16

**Comment by robinpellegrims** [comment]
> looks like `e-square-io/nx-distributed-task@v2` is expecting `@nrwl/cli` to be available but with this update that scope changes to `@nx/cli` we might need an update in that repo as well with a major bump supporting nx 16

not sure that's the cause, it seems unable to find the nx binary, but that one should still be there.. The install reports this as well: `NX   Failed to parse pnpm lockfile`

I'll take a deeper look when I find some time.

**Comment by rathpc** [comment]
> > looks like `e-square-io/nx-distributed-task@v2` is expecting `@nrwl/cli` to be available but with this update that scope changes to `@nx/cli` we might need an update in that repo as well with a major bump supporting nx 16
> 
> not sure that's the cause, it seems unable to find the nx binary, but that one should still be there.. The install reports this as well: `NX   Failed to parse pnpm lockfile`
> 
> I'll take a deeper look when I find some time.
> 

Oh interesting, I missed that failure

**Comment by rathpc** [comment]
> > looks like `e-square-io/nx-distributed-task@v2` is expecting `@nrwl/cli` to be available but with this update that scope changes to `@nx/cli` we might need an update in that repo as well with a major bump supporting nx 16
> 
> not sure that's the cause, it seems unable to find the nx binary, but that one should still be there.. The install reports this as well: `NX Failed to parse pnpm lockfile`
> 
> I'll take a deeper look when I find some time.

I was digging into this more - I am fairly confident it is due to the action being pretty outdated (nx v13.x.x). The error message actually comes directly from this action.

As for the postinstall error that is a known issue reporting in nx's issues and it is intended to be a soft failure and not actually related to this issue. 

I am trying to see how difficult it will be to update those github actions to v16 now

**Comment by rathpc** [comment]
> > > > looks like `e-square-io/nx-distributed-task@v2` is expecting `@nrwl/cli` to be available but with this update that scope changes to `@nx/cli` we might need an update in that repo as well with a major bump supporting nx 16
> > > 
> > > 
> > > not sure that's the cause, it seems unable to find the nx binary, but that one should still be there.. The install reports this as well: `NX Failed to parse pnpm lockfile`
> > > I'll take a deeper look when I find some time.
> > 
> > 
> > I was digging into this more - I am fairly confident it is due to the action being pretty outdated (nx v13.x.x). The error message actually comes directly from this action.
> > As for the postinstall error that is a known issue reporting in nx's issues and it is intended to be a soft failure and not actually related to this issue.
> > I am trying to see how difficult it will be to update those github actions to v16 now
> 
> oh ok, yeah I remember it was pretty difficult already to update to nx 15 on that github action, so not sure there will be a solution soon.. A temporary solution could be to postpone the @nrwl -> @nx replacements (while still updating to nx 16), since iirc nx 16 is backwards compatible with the @nrwl packages?

Yea it is supposed to be backwards compatible however because of the migrations pretty much everything is expected to change to the new scope and what I have seen is that if you have any dependency asking for the old one in the new it will fail. I suspect the same would happen the other direction as well.

The actions definitely will not be a simple lift but anyone using them hoping to upgrade to nx 16 will hit this issue. There are two paths I think. Add support for Nx 16 (in a very hacky way) or properly update the repo to nx 16 dropping support for previous versions and release a new major bump. I should have time today/tomorrow that I can try working on the latter but if you would want to attempt the hacky solution in the meantime it looks like there was already some hacky stuff added around v14 that we can follow for v16, expecially with regards to the binary but I don't know how deep the rabbit hole will go yet

**Comment by rathpc** [comment]
Added logic to hopefully get this working. https://github.com/e-square-io/nx-github-actions/pull/70

The changes needed to actually upgrade to v16 are pretty extensive and I think a lot of the code there would actually be duplication at that point as well. It may actually be simpler to rewrite most of the actions from scratch but that's for another day I suppose

**Comment by rathpc** [comment]
Hey again! https://github.com/e-square-io/nx-github-actions/pull/70 was merged so I believe you just need to rerun the workflow and it should pick up the latest v2 version.

**Comment by robinpellegrims** [comment]
In the meantime I changed the default pnpm version on our ci workflow to v8.5.1, which seems to be incompatible with the lockfile on this PR. Could you regenerate it with the latest version of pnpm?

**Comment by rathpc** [comment]
> In the meantime I changed the default pnpm version on our ci workflow to v8.5.1, which seems to be incompatible with the lockfile on this PR. Could you regenerate it with the latest version of pnpm?

Ah ok, yes I will push the change a little later today

**Comment by rathpc** [comment]
Pushed the change but the workflow is still failing because the actual change in the action hasn't been pushed to the project repo yet https://github.com/e-square-io/nx-distributed-task after my merge happened

**Comment by PatrickMennen** [comment]
Any estimate on when this will be merged (and more importantly) published to npm? the s3 cache plugin is great, but it is currently blocking us from going to nx 16.

**Comment by robinpellegrims** [comment]
> Any estimate on when this will be merged (and more importantly) published to npm? the s3 cache plugin is great, but it is currently blocking us from going to nx 16.

We're waiting for a new release of [e-square-io/nx-distributed-task](https://github.com/e-square-io/nx-distributed-task), but it should be ready any day now. See https://github.com/e-square-io/nx-github-actions/pull/70 for more info.

**Comment by rathpc** [comment]
@robinpellegrims so it looks like the only reason things are failing now are actually due to this project itself referencing an older version of its own runner. Essentially a side effect of needing a major bump update in its own dependencies before it actually exists. I don't know how you would feel about ignoring the fact things are failing in order to deploy a new major version? Once that exists I can update the dependency and open a patch PR to use the proper runner internally

**Comment by robinpellegrims** [comment]
> @robinpellegrims so it looks like the only reason things are failing now are actually due to this project itself referencing an older version of its own runner. Essentially a side effect of needing a major bump update in its own dependencies before it actually exists. I don't know how you would feel about ignoring the fact things are failing in order to deploy a new major version? Once that exists I can update the dependency and open a patch PR to use the proper runner internally

Yeah, I think you're right it would be more interesting to reference the current version of the library locally in this repository, rather than referencing the published version. We can change that later on. For now I'll merge the PR on master so we can move on.

**Comment by rathpc** [comment]
> > @robinpellegrims so it looks like the only reason things are failing now are actually due to this project itself referencing an older version of its own runner. Essentially a side effect of needing a major bump update in its own dependencies before it actually exists. I don't know how you would feel about ignoring the fact things are failing in order to deploy a new major version? Once that exists I can update the dependency and open a patch PR to use the proper runner internally
> 
> Yeah, I think you're right it would be more interesting to reference the current version of the library locally in this repository, rather than referencing the published version. We can change that later on. For now I'll merge the PR on master so we can move on.

Ok, I will open another PR. I think this one should have triggered a major bump but I am unfamiliar with your release process. (it looks like it released under the v2 major instead)

**Comment by robinpellegrims** [comment]
Yes I think the reason it didn't do a major bump is because the breaking change commit was typed 'chore' instead of 'feat', and chore commits are ignored during release version calculation. Probably on the next feat commit it will do a major bump though. You can check by doing a dry run locally using `nx release nx-remotecache-s3` and see what version it calculates.

**Pull request merged**
